src/waiter.c

```diff
@@ -0,0 +1,311 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include <stdalign.h>
+#include <stdbool.h>
+#include <inttypes.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/thandle.h"
+#include "c_pal/sm.h"
+#include "c_pal/srw_lock.h"
+
+#include "c_util/async_op.h"
+#include "c_util/rc_ptr.h"
+
+#include "c_util/waiter.h"
+
+typedef struct WAITER_TAG
+{
+    SM_HANDLE sm;
+    SRW_LOCK_HANDLE lock;
+    THANDLE(RC_PTR) data;
+    NOTIFICATION_CALLBACK notification_callback;
+    void* notification_callback_context;
+    NOTIFY_COMPLETE_CALLBACK notify_complete_callback;
+    void* notify_complete_callback_context;
+    bool registered;
+    bool notified;
+    bool callbacks_called;
+} WAITER;
+
+WAITER_HANDLE waiter_create(void)
+{
+    WAITER_HANDLE result;
+    WAITER_HANDLE waiter = malloc(sizeof(WAITER));
+    if (waiter == NULL)
+    {
+        LogError("Failure in malloc(sizeof(WAITER_HANDLE))");
+        result = NULL;
+    }
+    else
+    {
+        waiter->sm = sm_create("waiter");
+        if (waiter->sm == NULL)
+        {
+            LogError("Failure in sm_create(\"waiter\")");
+            result = NULL;
+        }
+        else
+        {
+            SM_RESULT sm_open_begin_result = sm_open_begin(waiter->sm);
+            if(sm_open_begin_result != SM_EXEC_GRANTED)
+            {
+                LogError("Failure in sm_open_begin(waiter->sm)");
+                result = NULL;
+            }
+            else
+            {
+                sm_open_end(waiter->sm, true);
+                waiter->lock = srw_lock_create("waiter", false);
+                if (waiter->lock == NULL)
+                {
+                    LogError("Failure in srw_lock_create(\"waiter\", false)");
+                    result = NULL;
+                }
+                else
+                {
+                    THANDLE_INITIALIZE(RC_PTR)(&waiter->data, NULL);
+                    waiter->notification_callback = NULL;
+                    waiter->notification_callback_context = NULL;
+                    waiter->notify_complete_callback = NULL;
+                    waiter->notify_complete_callback_context = NULL;
+                    waiter->registered = false;
+                    waiter->notified = false;
+                    waiter->callbacks_called = false;
+                    result = waiter;
+                    goto all_ok;
+                }
+            }
+            sm_destroy(waiter->sm);
+        }
+        free(waiter);
+    }
+all_ok:
+    return result;
+}
+
+static void abandon(void* context)
+{
+    WAITER_HANDLE waiter = (WAITER_HANDLE)context;
+    srw_lock_acquire_exclusive(waiter->lock);
```

Comment:
>lock [](http://example.com/codeflow?start=39&length=4)

Why does abandon even need this lock since abandon is only called from _destroy, and _destroy is never supposed to be thread-safe in our architecture.

Reply:
removing lock from abandon